Handle avatar image error

MidhunSureshR opened the pull request:
Fixes #402 
Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Handle avatar error

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/avatar.js:
```diff
@@ -65,6 +70,8 @@ export class AvatarView extends BaseUpdateView {
         this._avatarTitleChanged();
         this._root = renderStaticAvatar(this.value, this._size);
+        const image = this._root.firstChild;
+        image?.addEventListener("error", () => this.setAvatarError());
```
Perhaps also listen for load here and in both handlers, remove the listeners of both, so once an image is loaded or failed to load, there are no listeners anymore.

bwindels commented on src/platform/web/ui/avatar.js:
```diff
@@ -65,6 +70,8 @@ export class AvatarView extends BaseUpdateView {
         this._avatarLetterChanged();
         this._avatarTitleChanged();
         this._root = renderStaticAvatar(this.value, this._size);
+        const image = this._root.firstChild;
```
Can we also do this for static avatars (e.g. done with renderStaticAvatar and not binding to the view model)? We'll need to handle both separately though I think, as we need to know in update if the avatar failed to load to add or remove the background color class.

bwindels reviewed (commented):
Thanks! Looks mainly good, just a few small nits.

MidhunSureshR committed:
Add and remove opposing event listeners

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/avatar.js:
```diff
@@ -57,6 +57,20 @@ export class AvatarView extends BaseUpdateView {
+            const image = e.target;
+            image.removeEventListener("error", handleAvatarError);
+        };
+        image?.addEventListener("error", handleAvatarError);
```
don't you need to remove both handlers on either event? I think you can use the same handler for both events even, and set `this._avatarError` to `e.type === "error";`.

MidhunSureshR replied:
Oops! Sorry about that

MidhunSureshR committed:
remove both handlers

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Revert commits

This reverts commit 5f1346568de7ff814799ac43b1981e0ff3417b5f.

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
implement new approach

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Move AvatarView to separate file

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
use textContent

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/Platform.js:
```diff
@@ -189,6 +190,7 @@ export class Platform {
                 this._disposables.track(disposable);
             }
         }
+        this._container.addEventListener("error", handleAvatarError, true);
```
Can you add a closure to disposables to remove this again when dispose gets called? `adaptUIOnVisualViewportResize` has an example of this.

bwindels commented on src/platform/web/ui/avatar.js:
```diff
@@ -108,16 +25,29 @@ export function renderStaticAvatar(vm, size, extraClasses = undefined) {
     }
     const avatarContent = hasAvatar ? renderImg(vm, size) : text(vm.avatarLetter);
-    return tag.div({className: avatarClasses}, [avatarContent]);
+    return tag.div({className: avatarClasses, 'data-avatar-letter': vm.avatarLetter}, [avatarContent]);
```
You could only set the letter if `hasAvatar` is true

bwindels commented on src/platform/web/ui/avatar.js:
```diff
@@ -108,16 +25,29 @@ export function renderStaticAvatar(vm, size, extraClasses = undefined) {
+    if (!isAvatarEvent(e)) { return; }
+    const parent = e.target.parentElement;
+    const avatarLetter = parent.getAttribute("data-avatar-letter");
+    parent.textContent = avatarLetter;
```
Wouldn't you also need to hide the img somehow? (and unhide if the binding in AvatarView changes the `src` property)

bwindels replied:
textContent already clears the image

bwindels commented on src/platform/web/ui/AvatarView.js:
```diff
@@ -0,0 +1,84 @@
+            const bgColorClass = `usercolor${vm.avatarColorNumber}`;
+            if (vm.avatarUrl(this._size)) {
+                this._root.replaceChild(renderImg(vm, this._size), this._root.firstChild);
+                this._root.classList.remove(bgColorClass);
```
The background color class should not be removed or set here I imagine, but just always left as it was by the initial render in `renderStaticAvatar`. You would however need to add/remove a class to show or hide the image when it fails to load I think?

bwindels reviewed (commented):
Looking good, just a few more things.

MidhunSureshR committed:
Pass color as data attribute

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
Only add attribute if we have avatar

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

MidhunSureshR committed:
remove listener on dispose

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels commented on src/platform/web/ui/AvatarView.js:
```diff
@@ -0,0 +1,84 @@
+            img.setAttribute("title", vm.avatarTitle);
+        }
+        if (this._avatarLetterChanged() && !hasAvatar) {
+            this._root.firstChild.textContent = vm.avatarLetter;
```
Can you do `this._root.textContent = ...` here as well like you did in handleAvatarError?

bwindels commented on src/platform/web/ui/AvatarView.js:
```diff
@@ -0,0 +1,84 @@
+        }
+        const hasAvatar = !!vm.avatarUrl(this._size);
+        if (this._avatarTitleChanged() && hasAvatar) {
+            const img = this._root.firstChild;
```
Can you check here now if this is indeed an image? As we might have removed it on error ...

bwindels commented on src/platform/web/ui/AvatarView.js:
```diff
@@ -0,0 +1,84 @@
+                this._root.replaceChild(renderImg(vm, this._size), this._root.firstChild);
+                this._root.classList.remove(bgColorClass);
+            } else {
+                this._root.replaceChild(text(vm.avatarLetter), this._root.firstChild);
```
Can you do `this._root.textContent = ...` here as well like you did in handleAvatarError?

bwindels commented on src/platform/web/ui/avatar.js:
```diff
@@ -108,16 +25,36 @@ export function renderStaticAvatar(vm, size, extraClasses = undefined) {
+    const parent = e.target.parentElement;
+    const avatarColorNumber = parent.getAttribute("data-avatar-color");
+    parent.classList.add(`usercolor${avatarColorNumber}`);
+    const avatarLetter = parent.getAttribute("data-avatar-letter");
```
actually, looks like you want to use innerText to make the <img> go away in IE11
https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent#differences_from_innertext

MidhunSureshR committed:
Make changes

- use textContent where possible
- make sure we have an image before adding title

Signed-off-by: RMidhunSuresh <rmidhunsuresh@gmail.com>

bwindels reviewed (approved):
Looks good, thanks! Did you test that it works as well?
